Platform/ElkhartlakeBoardPkg/Library/Stage2BoardInitLib/Stage2BoardInitLib.c

```diff
@@ -591,39 +591,30 @@ UpdatePayloadId (
   )
 {
   EFI_STATUS      Status;
-  UINT32          PayloadSelGpioData;
-  UINT32          PayloadSelGpioPad;
   GEN_CFG_DATA    *GenCfgData;
+  UINT32          PayloadId;
+  UINT32          PayloadSelGpioData;
+  UINT32          PayloadSelGpioPin;
 
   GenCfgData = (GEN_CFG_DATA *)FindConfigDataByTag (CDATA_GEN_TAG);
   if (GenCfgData == NULL) {
     ASSERT (FALSE);
     return;
   }
-  SetPayloadId (GenCfgData->PayloadId);
 
-  if (GetPayloadId () != AUTO_PAYLOAD_ID_SIGNATURE) {
+  if (GenCfgData->PayloadId != AUTO_PAYLOAD_ID_SIGNATURE || !GenCfgData->PayloadSelGpio.Enable) {

```

Comment:
This condition should be AND instead of OR?
Should we set default payload if invalid option was set? (eg. GenCfgData->PayloadId = 'UEFI' and GenCfgData->PayloadSelGpio.Enable = 1)?

Reply:
it has to be 'or' condition.
if the uefi payload is set cfgdata (payload id) = uefi, then it will set as per cfgdata. 
but if it is set as auto detection while gpio pin selection is not set, then it will still use the default value (null), which is osloader.